Reimplement BigQuery query execution to use GetQueryResults

jskeet opened the pull request:
Background: we used to use GetQueryResults, then we were asked to use ListRows - but this slows down queries, as observed in #2876. It turns out that GetQueryResults is now the preferred client library usage, so it's worth changing back to it.

(This review is fine to wait until you're back from the US, Chris.)

chrisdunelm commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryResults.cs:
```diff
@@ -122,10 +163,32 @@ public BigQueryResults(BigQueryClient client, GetQueryResultsResponse response,
+
+            // Work out whether to use the response we've already got, or create a new one.
+            GetQueryResultsResponse response = _response;
+            if ((clonedOptions.PageSize == null || clonedOptions.PageSize > pageSize) && _response.Rows?.Count > pageSize)
```
I'm unclear why this isn't just `if (_response.Rows?.Count > pageSize)`. If `_response` contains more than `pageSize` rows, doesn't it always need re-starting?

jskeet replied:
Yes, I think you're right. Will simplify it in both places.

chrisdunelm commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryResults.cs:
```diff
@@ -192,5 +260,58 @@ public BigQueryResults ThrowOnAnyError()
+        {
+            private readonly GetQueryResultsOptions _options;
+            private readonly BigQueryResults _parent;
+            private IEnumerator<BigQueryRow> _underlyingIterator;
```
nit: rename to `_underlyingEnumerator`?

jskeet replied:
I guess. I prefer the term "iterator" over "enumerator", but that's just a personal thing. Done.

chrisdunelm commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryResults.cs:
```diff
@@ -192,5 +260,58 @@ public BigQueryResults ThrowOnAnyError()
+
+            public async Task<bool> MoveNext(CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```
Should this be inside the `while` loop below?

jskeet replied:
We at least need one call *outside* the loop, so that it can still be cancelled even if we're iterating over the content of an existing response. Within the loop, I don't *think* we need one as I'd expect the cancellation token usage in the RPC to be enough.

chrisdunelm replied:
Ah yes, I had missed that `cancellationToken` is passed into `GetRawQueryResultsAsync(...)`.

jskeet commented:
Added another commit addressing the comments, but would appreciate a look at them if you have time - particularly my reasoning around *not* changing the cancellation check placement.
